Show dialog if owners field set to be empty on shiro

soralee opened the pull request:
### What is this PR for?
This PR is for a showing dialog when note's permission set `Readers/Writers` but `Owners` is empty. 
Currently, after merged #849, if user set `Owners` is empty, it's automatically set as the current user.
So, It would be better to show a dialog before `Owners` is fill with the current user because user might think it is not intended. 

Plus, if a user only set `Readers` role, all field(`Writers` and `Owners`) role is filled as the current user like below. It's not necessary to fill because the `Owner` role can read and write note.
![not_need_wirters](https://cloud.githubusercontent.com/assets/8110458/24869746/2a77d2da-1e4f-11e7-86a5-794e1aac02d0.gif)


### What type of PR is it?
[ Improvement ]


### What is the Jira issue?
* [ZEPPELIN-2206; Be added a writer account when adding a reader account on Shiro](https://issues.apache.org/jira/browse/ZEPPELIN-2206)
* [ZEPPELIN-2418; show dialog if user set to the empty owner permission](https://issues.apache.org/jira/browse/ZEPPELIN-2418)

### How should this be tested?
1. Turn on shiro.
2. Set permission `Writers` and `Readers` except `Owner` role.
3. Check the dialog appears

### Screenshots (if appropriate)
[Before]
* Set `Readers` and `Writers` except `Owners`
![zeppelin-2206_auto_shiro_b_1](https://cloud.githubusercontent.com/assets/8110458/24869964/cf382676-1e4f-11e7-8e95-30c149a4aba5.gif)

* Set only 'Writers`
![zeppelin-2206_auto_shiro_b_2](https://cloud.githubusercontent.com/assets/8110458/24870112/3821f518-1e50-11e7-9310-fcafd4e26376.gif)

* set only `Readers`; this screen is attached above.

[After]
* Set `Readers` and `Writers` except `Owners` ( Set only `Writers`; this result is same)
![zeppelin-2206_auto_shiro_a_1](https://cloud.githubusercontent.com/assets/8110458/24870781/61d8f1c0-1e52-11e7-8faa-38ff4fd64f6f.gif)

* Set only 'Readers`
![zeppelin-2206_auto_shiro_a_2](https://cloud.githubusercontent.com/assets/8110458/24870811/76f547ca-1e52-11e7-98b2-f6a5541284ab.gif)


### Questions:
* Does the licenses files need update? No
* Is there breaking changes for older versions? No
* Does this needs documentation? No

soralee commented:
@felixcheung Thanks for review! 
Yeah, you're right. It would be better if we also check on the REST API side stably.
So, let me create a JIRA ticket and revert code in `REST API`.

soralee commented:
Hi, @felixcheung :slightly_smiling_face: 
I added JIRA url about this improvement in this PR description and I reverted some code in `notebook Rest API` side except that, if only setting `Readers` role, all field(Writers and Owners) role is filled as the current user. Because the `Owners` role can already read and write note, we don't need to fill in `Writers` role.

felixcheung commented on zeppelin-web/src/app/notebook/notebook.controller.js:
```diff
@@ -750,6 +750,35 @@ function NotebookCtrl($scope, $route, $routeParams, $location, $rootScope,
+    if ($scope.isOwnerEmpty()) {
+      BootstrapDialog.show({
+        closable: false,
+        title: 'warning Permissions',
```
it's a bit odd if this is the display title of the dialog `warning Permissions`?
perhaps just `Setting Owners Permissions`?

felixcheung commented on zeppelin-web/src/app/notebook/notebook.controller.js:
```diff
@@ -750,6 +750,35 @@ function NotebookCtrl($scope, $route, $routeParams, $location, $rootScope,
+          'Current user : [ ' + $rootScope.ticket.principal + ']',
+        buttons: [
+          {
+            label: 'OK',
```
instead of having an `OK` button to set/change permission, how about having it as `Set` to be more clear?

felixcheung commented on zeppelin-web/src/app/notebook/notebook.controller.js:
```diff
@@ -861,6 +890,14 @@ function NotebookCtrl($scope, $route, $routeParams, $location, $rootScope,
   };
 
+  $scope.isOwnerEmpty = function() {
+    if ($scope.permissions.owners.length > 0){
```
should we check if it's just whitespace? `       `

1ambda commented:
Can we remove `!!!` from dialog title? it's meaningless and noisy. although it's not @soralee's change, i think we can handle it in this PR.

soralee commented:
@felixcheung Thanks for detail suggestion! That's sounds good. Let me update as each comment.
@1ambda Sure, I agree! I'll remove `!!!`

1ambda commented:
@soralee Since conflicts are about lint errors, you can simply rebase by

1. keeping your change and run `rebase --continue`
2. then, execute  `eslint src --fix` in `zeppelin-web` dir.

soralee commented:
@1ambda Thanks for information always :+1: 
I already worked to rebase and run build successfully, but now, I'm checking the `white space` problem.
Let me updated ASAP!

soralee commented:
I modified dialog title and `OK` button.
I've checked the `whitespace` and the `$scope.permissions.owners` array don't insert `whitespace` originally. But for stability, I added `if statement`. 
@felixcheung, @1ambda  could you guys review this again, please?

felixcheung commented on zeppelin-web/src/app/notebook/notebook.controller.js:
```diff
@@ -867,6 +896,20 @@ function NotebookCtrl ($scope, $route, $routeParams, $location, $rootScope,
+  $scope.isOwnerEmpty = function() {
+    if ($scope.permissions.owners.length > 0) {
+      for (let i = 0; i < $scope.permissions.owners.length; i++) {
+        $scope.permissions.owners[i] = $scope.permissions.owners[i].trim()
```
check `$scope.permissions.owners[i].trim()` doesn't return empty string before setting it?

felixcheung replied:
and what if we have to skip some owners before it's empty/whitespace - might wan to copy to a new list / keep the left hand side index separately?

soralee replied:
@felixcheung Thanks for review and useful comments!
You're right! I initially think something wrong. Let me modify this part again!
Thanks a lots!

felixcheung commented on zeppelin-web/src/app/notebook/notebook.controller.js:
```diff
@@ -752,14 +752,43 @@ function NotebookCtrl ($scope, $route, $routeParams, $location, $rootScope,
+        closable: false,
+        title: 'Setting Owners Permissions',
+        message: 'Please fill the [Owners] field. If not, it will set as current user.\n\n' +
+          'Current user : [ ' + $rootScope.ticket.principal + ']',
```
not sure if we need to check $rootScope.ticket.principal  is empty or whitespace

soralee replied:
@felixcheung thanks for carefully checking this.
I didn't care this `$rootScope.ticket.principal` variable, because this `$scope.savepermissions` function is not run when user doesn't use Siro and user didn't login. 
But, for prevent, I would be better to add `if statement`for checking about this case that `$rootScope.ticket.principal` is ayonymous or empty.

soralee commented:
@felixcheung Thanks for review and useful comments!
I initially think something wrong. Let me modify this part again!
Thanks a lots!

soralee commented:
@khalidhuseynov Thanks for comments!
You mean, this can be occurred another problem like the following steps. 
1. A user who has admin permissions sets `readers` and `writers` and empty `owners` roles in `Note permissions'.
2. The user clicks 'Save' button in `Note permissions`
3. The user clicks `Cancel button` in the new dialog(`Setting Owners Permissions`) which I made.
3. then, the default `owners` value is disappeared.
4. and, the user add `user1` to `owners` role and click 'Save' button.

As my understanding your comment, this 3 and 4 steps can be a problem. right?
@khalidhuseynov, Could you check what I understood correctly, please?

khalidhuseynov commented:
@soralee yeah the steps are correct. so if you follow those steps and click `Save`, say for `Note A`, then `Note A` will disappear from the list of your notes, and then you will not have permissions to view `Note A`

soralee commented:
@khalidhuseynov thanks for checking it really!
You're right! I think so too! this should be considered.
Furthermore, I checked above steps which I addressed, I saw that zeppelin run strange operation.  
Let me improve it so that it does not disappear.

khalidhuseynov commented:
@soralee it's disappearance is based on multi-user logic and it's correct since if note have `owners`, `readers`, `writers` as `user1` it doesn't have anything to do in `admin`'s list. That's why I was telling that not sure that this PR makes sense, because of intervening into that flow. From another point when creating note, it's `owner` is automatically assigned as current user without asking as well, so I think having already existing logic and assigning current user as owner (unless having admin feature) is fine, and we don't need much changes on it for now

soralee commented:
@khalidhuseynov Thanks for opinion! Yeah, I see. This PR is for showing a dialog when (already existing backend logic) that note's permission set Readers/Writers but Owners is empty. I think I would better to focus the purpose! 
So, next step is that I'd like to update this PR as @felixcheung addressed.

soralee commented:
@felixcheung I updated as your advice!
@khalidhuseynov I added a logic which is reverted with old value if user cancel button like above `4 step`. How about this?

> 4. then, the default owners value is disappeared.

![revert_shiro](https://cloud.githubusercontent.com/assets/8110458/25418845/bfd841d4-2a89-11e7-97d7-63e7072d0b99.gif)

soralee commented:
Ping!
And this CI fail is irrelevant this PR.
```
Tests in error: 
  SecurityRestApiTest.testGetUserList:69 » JsonSyntax java.lang.IllegalStateExce...
  SecurityRestApiTest.testTicket:55 » JsonSyntax java.lang.IllegalStateException...
Tests run: 58, Failures: 5, Errors: 2, Skipped: 1
```

soralee commented:
Hi, there.
CI became green now.

soralee commented:
Ping!
@felixcheung @khalidhuseynov @1ambda, could you guys confirm again, please?

soralee commented:
Ready to review!

felixcheung commented on zeppelin-server/src/main/java/org/apache/zeppelin/rest/NotebookRestApi.java:
```diff
@@ -208,9 +208,6 @@ public Response putNotePermissions(@PathParam("noteId") String noteId, String re
     // Set readers, if writers and owners is empty -> set to user requesting the change
     if (readers != null && !readers.isEmpty()) {
-      if (writers.isEmpty()) {
-        writers = Sets.newHashSet(SecurityUtils.getPrincipal());
```
I might have missed this, why are we making this change?

soralee replied:
Thanks for good point!
This line `if` statement mean that when filling `readers` and not filling `writers`, it's filled with `writers` as current user. In this case, I think, it's enough for filling with `owners` as current user. 
And, the following screenshot and explain is in this PR description.
> Plus, if a user only set Readers role, all field(Writers and Owners) role is filled as the current user like below. It's not necessary to fill because the Owner role can read and write note.
https://cloud.githubusercontent.com/assets/8110458/24869746/2a77d2da-1e4f-11e7-86a5-794e1aac02d0.gif

soralee commented:
@felixcheung, regarding your comment 
> hey sorry, so in your screen shot here #2239 (comment)
if cancel when owner is empty it resets also writer/reader, is that a bit odd?

that screenshot means it is reverted with old value if user click `cancel` button.
So, it's describing that 
 1. original data is only `owners: admin`, 
 2. and change data of roles
 3. and then click `cancel` button.
 4. make sure data is reverted by original data (`owners: admin`)

soralee commented:
reopen for CI

khalidhuseynov commented:
Merging into master if no more discussion
